[SPARK-19956][Core]Optimize a location order of blocks with topology information

ConeyLiu opened the pull request:
## What changes were proposed in this pull request?

When call the method getLocations of BlockManager, we only compare the data block host. Random selection for non-local data blocks, this may cause the selected data block to be in a different rack. So in this patch to increase the sort of the rack.

## How was this patch tested?

New test case.

Please review http://spark.apache.org/contributing.html before opening a pull request.

ConeyLiu committed:
optimize a location order of blocks with topology information

AmplabJenkins commented:
Can one of the admins verify this patch?

jerryshao commented:
The fix LGTM, I think it is nice to have such topology priority. CC @mridulm .

mridulm commented on core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:
```diff
@@ -500,6 +500,30 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
+    val getLocations = PrivateMethod[Seq[BlockManagerId]]('getLocations)
+    val locations = blockManager invokePrivate getLocations(BroadcastBlockId(0))
+    assert(locations.map(_.host).toSet
+      === Set(localHost, localHost, otherHost, otherHost, otherHost))
```
Remove `toSet` and make `Set` as `Seq` ?
Making it a Set sort of looses the point of this change (ordering is lost).
Same for topology info below below.

mridulm commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -555,12 +555,15 @@ private[spark] class BlockManager(
   /**
    * Return a list of locations for the given block, prioritizing the local machine since
-   * multiple block managers can share the same host.
+   * multiple block managers can share the same host, then try to get the same rack data.
```
How about "then try to get the same rack data" -> "followed by hosts on the same rack" ?

ConeyLiu committed:
Address comments

ConeyLiu commented:
Hi, @jerryshao @mridulm Thanks for your review, I have updated the code.

ConeyLiu commented on core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala:
```diff
@@ -497,7 +497,30 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
     val getLocations = PrivateMethod[Seq[BlockManagerId]]('getLocations)
     val locations = blockManager invokePrivate getLocations(BroadcastBlockId(0))
-    assert(locations.map(_.host).toSet === Set(localHost, localHost, otherHost))
+    assert(locations.map(_.host) === Seq(localHost, localHost, otherHost))
```
There should also be Seq.

mridulm reviewed (approved):
Looks good to me, will keep it open in case there are other comments.
+CC @jerryshao , @ConeyLiu

ConeyLiu commented:
Ok, thanks a lot.

ConeyLiu commented:
Hi, @cloud-fan @zsxwing Can you take a look? Thanks a lot.

cloud-fan commented:
retest this please

cloud-fan commented:
LGTM

SparkQA commented:
**[Test build #76453 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76453/testReport)** for PR 17300 at commit [`56f5231`](https://github.com/apache/spark/commit/56f5231626cceb114c45413b7b340ee719c3f2f8).

SparkQA commented:
**[Test build #76453 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76453/testReport)** for PR 17300 at commit [`56f5231`](https://github.com/apache/spark/commit/56f5231626cceb114c45413b7b340ee719c3f2f8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76453/
Test FAILed.

ConeyLiu commented:
retest this please

mridulm commented:
Will merge when tests pass.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #76473 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76473/testReport)** for PR 17300 at commit [`56f5231`](https://github.com/apache/spark/commit/56f5231626cceb114c45413b7b340ee719c3f2f8).

viirya commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -555,12 +555,15 @@ private[spark] class BlockManager(
     val (preferredLocs, otherLocs) = locs.partition { loc => blockManagerId.host == loc.host }
-    preferredLocs ++ otherLocs
+    val (sameRackLocs, differentRackLocs) = otherLocs.partition {
+      loc => blockManagerId.topologyInfo == loc.topologyInfo
```
If `blockManagerId.topologyInfo` is `None`, we will prefer the locations with empty `topologyInfo`. It is slightly different with what the shuffling wants to do here.

ConeyLiu replied:
Modified, thanks a lot for the good advice.

ConeyLiu committed:
Merge branch 'master' into blockmanager

ConeyLiu committed:
address comments, and fix the test error

ConeyLiu commented:
Thanks both of you for review, I have addressed the comments and modified the test case. Please help calling jenkins for test, because I can't trigger that.  Thanks again.

SparkQA commented:
**[Test build #76473 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76473/testReport)** for PR 17300 at commit [`56f5231`](https://github.com/apache/spark/commit/56f5231626cceb114c45413b7b340ee719c3f2f8).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76473/
Test FAILed.

mridulm commented:
retest this please

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #76500 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76500/testReport)** for PR 17300 at commit [`a3e979f`](https://github.com/apache/spark/commit/a3e979f5e421b7c5d07811eee001f5b086166093).

mridulm commented:
That is weird, jenkins should have restarted build when I commented ...

SparkQA commented:
**[Test build #76500 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76500/testReport)** for PR 17300 at commit [`a3e979f`](https://github.com/apache/spark/commit/a3e979f5e421b7c5d07811eee001f5b086166093).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76500/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #76502 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76502/testReport)** for PR 17300 at commit [`a3e979f`](https://github.com/apache/spark/commit/a3e979f5e421b7c5d07811eee001f5b086166093).

SparkQA commented:
**[Test build #76502 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76502/testReport)** for PR 17300 at commit [`a3e979f`](https://github.com/apache/spark/commit/a3e979f5e421b7c5d07811eee001f5b086166093).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76502/
Test PASSed.

viirya commented:
LGTM

ConeyLiu commented:
Thanks for your review.

cloud-fan commented:
thanks, merging to master!

mridulm commented:
Thanks for merging @cloud-fan, this PR kept dropping form my list ...

ConeyLiu commented:
Thanks @cloud-fan @mridulm @gatorsmile
